### DISCUSSION TITLE
How to get storageKeys for storage subscription?

### DISCUSSION THREAD

**Discussion by poka-IT** [rankanizer/polkadart#443] [Q&A] [answered]
How to use `subscribeStorage` ? I don't find examples.
I can find the storageKeys on polakdot.js UI, but I don't understand how to get it from this lib?

How to subscribe new blocks, or subscribe to account balance for instance?

```
duniterApi.rpc.state.subscribeStorage(storageKeysForAccountBalance, (data) {
  print(data.data.free);
});
```

Thank you!

**Comment by leonardocustodio** [comment]
You can refer to a sample here: https://github.com/leonardocustodio/polkadart/issues/385#issuecomment-1870540334
But just so you know, may you ever need to subscribe to a storage key; besides being able to look at the `Chain State` tab on Polkadot-JS, the `encodedStorageKey` shall you ever need to encode it, it is basically:
```
Twox128(pallet) + Twox128(method)
```
Anyway, those should also be available through the lib as in the example above:

![image](https://github.com/leonardocustodio/polkadart/assets/5619696/a8f623ab-fa06-489b-82d3-8c932b6820ad)

**Comment by poka-IT** [comment, marked as answer]
Thank you, his example works :)

So this is my test code working:

```dart
Future<void> startBlockListening() async {
  final storageBlockKeyQuery = [duniterApi.query.system.numberKey()];
  await duniterApi.rpc.state.subscribeStorage(
    storageBlockKeyQuery,
    (storageChangeSet) async {
      final blockNumber = await duniterApi.query.system.number();
      log.d(blockNumber);
    },
  );
}

final Map<String, StreamSubscription<StorageChangeSet>>
    _addressSubscriptionMap = {};

Future<void> startBalanceListening(String address) async {
  if (_addressSubscriptionMap[address] != null) return;
  final accountPubKey = Address.decode(address).pubkey;
  final storageAccountKeyQuery = [
    duniterApi.query.system.accountKey(accountPubKey)
  ];
  _addressSubscriptionMap[address] =
      await duniterApi.rpc.state.subscribeStorage(
    storageAccountKeyQuery,
    (storageChangeSet) async {
      final accountInfo =
          await duniterApi.query.system.account(accountPubKey);
      log.d(accountInfo.data.free);
    },
  );
}

Future<void> stopBalancesListening(String address) async {
  if (_addressSubscriptionMap[address] == null) return;
  final subscription = _addressSubscriptionMap[address];
  if (subscription == null) return;
  _addressSubscriptionMap.remove(address);
  await subscription.cancel();
}
```

But the thing is, here we resend a request for each changes:
`final blockNumber = await duniterApi.query.system.number();`

Isn't there a way to avoid this by retrieving the correct data directly from `storageChangeSet`?

**Comment by leonardocustodio** [comment]
The best way to subscribe to `blocks` is not to subscribe directly to the storage but to use the RPC.
There are `chain_subscribeNewHeads` and `chain_subscribeFinalizedHeads`. You can use either one.
If  you are wondering how to do that on Polkadart, you can use the `Provider` class, initialize it, and call `subscribe` using the method/args you want.
